src/osd/OSD.cc

```diff
@@ -365,83 +366,94 @@ void OSDService::identify_splits_and_merges(
     queue.pop_front();
     did.insert(cur);
     unsigned pgnum = old_pgnum;
-    for (auto q = p->second.lower_bound(old_map->get_epoch());
-	 q != p->second.end() &&
-	   q->first <= new_map->get_epoch();
-	 ++q) {
-      if (pgnum < q->second) {
+    // Note: the new_epoch in which the pg_num change was recorded is *not*
+    //       the exact epoch in which the actual change was made.
+    //       The iteration starts from old_map's lower bound epoch up to
+    //       new_map's epoch. Therfore, we don't need the actual epoch
+    //       of the pg num change - as long as it's recorded before the new_map.
+    //       The last statement is true because we record all of the
+    //       changes from old_map's epoch up to new_map's epoch.
+    //       See: OSD::track_pools_and_pg_num_changes
```

Comment:
This comment only really mentions the loop bound here, but there are other places those epochs are used.  Below, we actually populate split_children and merge_pgs using the epochs in the this pool_pg_num_history_map set.  Moreover, the implementation in this PR would actually allow us to miss splits or merges -- suppose the we receive a message with the following maps:

10: pg_num 16->32
11: pg_num 32->24
12: pg_num 24->64
13: no change

track_pools_and_pg_num_changes would only record

13: 64

resulting in *split_children containing (13: 64) and *merge_pgs being empty rather than

10: 32
11: 24
12: 64

with *split_children (10:32, 12:64) and *merge_pgs (11:24).